Add user deletion to new Account Management page

Amanda-Kauppinen opened the pull request:
This pull request introduces the capability for users to delete their own user accounts using the [Marketplace API current_user/delete endpoint](https://www.sharetribe.com/api-reference/marketplace.html#delete-user).

This pull request also introduces a new Account settings tab, Manage account. Currently, user deletion is the only feature on that tab.

<img width="923" height="675" alt="Screenshot 2025-11-11 at 12 58 42" src="https://github.com/user-attachments/assets/ee9d98ff-3538-4bd4-bf58-f11b0a57cbce" />

When a user checks the checkbox to indicate they want to delete their account, they are prompted to confirm the deletion with their password. The user can also request a password reset on the same page.

<img width="926" height="665" alt="Screenshot 2025-11-11 at 12 58 55" src="https://github.com/user-attachments/assets/b716b675-570b-410e-8fa7-e574b88a4104" />

Since the `current_user/delete` endpoint requires a trusted context, the request is routed through the template server using the `/api/delete-account` endpoint. The endpoint first queries the user's transactions to check whether there are transactions in incomplete payment states. If any such transactions exist, the endpoint throws an error to the client.

**Note:** When you are customizing your marketplace with custom transaction processes, you'll need to make sure that you reflect those changes in this part of the user deletion as well.

<img width="928" height="719" alt="Screenshot 2025-11-11 at 13 00 39" src="https://github.com/user-attachments/assets/240fe691-ed5c-46a4-93b2-9bc1ada370da" />

The `current_user/delete` endpoint has a new parameter `deleteFromStripe`. When this parameter is passed to the endpoint, the user's Stripe Connect account and Stripe Customer account are also deleted from Stripe if they exist. However, if the user's Stripe Connect account has a non-zero balance, we don't allow deleting the account, and users need to contact the marketplace operator to resolve the balance and delete their account.

<img width="944" height="732" alt="Screenshot 2025-11-11 at 13 56 59" src="https://github.com/user-attachments/assets/e35546e9-8434-4034-baca-b172185aae7e" />

## Translation changes
### New translations
```json
  "DeleteAccountForm.checkboxLabel": "I want to delete my account",
  "DeleteAccountForm.confirmDeleteInfo": "Deleting your {marketplaceName} account removes your personal data, user profile, and listings from the marketplace. This can't be undone.",
  "DeleteAccountForm.confirmDeleteTitle": "Confirm account deletion",
  "DeleteAccountForm.deleteAccount": "Delete account",
  "DeleteAccountForm.deleteAccountInfo": "To delete your account ({email}), please enter your current password.",
  "DeleteAccountForm.deleteAccountError": "Oops, something went wrong. Please try again.",
  "DeleteAccountForm.ongoingTransactionsError": "Account deletion failed. Please complete any ongoing transactions before deleting your account or contact support.",
  "DeleteAccountForm.passwordFailed": "Please double-check your password.",
  "DeleteAccountForm.passwordLabel": "Current password",
  "DeleteAccountForm.passwordPlaceholder": "Enter your current password…",
  "DeleteAccountForm.passwordTooShort": "The password should be at least {minLength} characters.",
  "DeleteAccountForm.passwordRequired": "A password is required.",
  "DeleteAccountForm.resendPasswordLinkText": "Resend instructions.",
  "DeleteAccountForm.resetPasswordInfo": "Forgot your password or don't have one? {resetPasswordLink}",
  "DeleteAccountForm.resetPasswordLinkSent": "Instructions to reset your password have been sent to {email}. {resendPasswordLink}",
  "DeleteAccountForm.resetPasswordLinkText": "Send reset instructions.",
  "DeleteAccountForm.stripeDeletionFailedError": "Account deletion failed. Stripe data could not be deleted. Please contact support.",
  "LayoutWrapperAccountSettingsSideNav.manageAccountTabTitle": "Manage account",
  "ManageAccountPage.heading": "Manage account",
  "ManageAccountPage.title": "Manage account",
```
### Updated translations

```diff
- "ModalMissingInformation.fixEmail": "Oops, typo in your email? {fixEmailLink}",
+ "ModalMissingInformation.fixEmail": "Oops, a typo in your email? {fixEmailLink}",
- "PasswordRecoveryPage.fixEmailInfo": "Oops, typo in your email? {fixEmailLink}",
+ "PasswordRecoveryPage.fixEmailInfo": "Oops, a typo in your email? {fixEmailLink}",
```

Amanda-Kauppinen committed:
Initial commit

Gnito commented on src/containers/DeleteAccountPage/DeleteAccountPage.duck.js:
```diff
@@ -0,0 +1,99 @@
+
+// ================ Action creators ================ //
+
+export const deleteAccountRequest = () => ({ type: DELETE_ACCOUNT_REQUEST });
+export const deleteAccountSuccess = () => ({ type: DELETE_ACCOUNT_SUCCESS });
```
You could follow "_changePasswordError_" and how it's used to show the error message on related form on PasswordChangePage.

Gnito commented on src/containers/DeleteAccountPage/DeleteAccountPage.js:
```diff
@@ -0,0 +1,162 @@
+    });
+  };
+
+  const deleteAccountForm = user.id ? (
+    <DeleteAccountForm
+      className={css.form}
+      initialValues={{ email: currentEmail }}
+      currentUser={currentUser}
+      onResendVerificationEmail={onResendVerificationEmail}
+      onResetPassword={onResetPassword}
+      onSubmit={handleSubmit}
+      inProgress={deleteAccountInProgress}
+      ready={accountDeletionConfirmed}
+      sendVerificationEmailInProgress={sendVerificationEmailInProgress}
+      sendVerificationEmailError={sendVerificationEmailError}
+      resetPasswordInProgress={resetPasswordInProgress}
+      resetPasswordError={resetPasswordError}
+    />
+  ) : null;
```
I suggest that you move this piece of ternary JSX inside the final return of rest of the JSX content.

```jsx
        <div className={css.content}>
          <H3 as="h1">
            <FormattedMessage id="DeleteAccountPage.heading" />
          </H3>
          {user.id ? (
            <DeleteAccountForm
              className={css.form}
              initialValues={{ email: currentEmail }}
              currentUser={currentUser}
              onResendVerificationEmail={onResendVerificationEmail}
              onResetPassword={onResetPassword}
              onSubmit={handleSubmit}
              inProgress={deleteAccountInProgress}
              ready={accountDeletionConfirmed}
              sendVerificationEmailInProgress={sendVerificationEmailInProgress}
              sendVerificationEmailError={sendVerificationEmailError}
              resetPasswordInProgress={resetPasswordInProgress}
              resetPasswordError={resetPasswordError}
            />
          ) : null}
        </div>

```
(variables are useful when they are used on multiple places, but if there's only single place where the form is included, then it just includes unnecessary back and forth jumping between file locations to understand what's happening)

Gnito commented on src/containers/DeleteAccountPage/DeleteAccountForm/DeleteAccountForm.module.css:
```diff
@@ -0,0 +1,146 @@
+  }
+}
+
+.phone {
```
Remember to clean this file so that it only contains classes that are actually in use.

Amanda-Kauppinen committed:
Rename DeleteAccountPage > ManageAccountPage

Amanda-Kauppinen committed:
ManageAccountPage: Moved DeleteAccountForm from variable assignment to the final return statement

Amanda-Kauppinen committed:
DeleteAccountForm: Cleaned up css file

Amanda-Kauppinen committed:
DeleteAccountForm: Converted to functional component (still very much WIP)

Gnito commented on src/translations/en.json:
```diff
@@ -479,6 +491,7 @@
   "LandingPage.schemaDescription": "An online marketplace.",
   "LandingPage.schemaTitle": "{marketplaceName}",
   "LayoutWrapperAccountSettingsSideNav.contactDetailsTabTitle": "Contact details",
+  "LayoutWrapperAccountSettingsSideNav.deleteAccountTabTitle": "Account management",
```
The key should be different
`LayoutWrapperAccountSettingsSideNav.manageAccountTabTitle`

Gnito commented on src/containers/ManageAccountPage/ManageAccountPage.js:
```diff
@@ -0,0 +1,148 @@
+          {user.id ? (
+            <DeleteAccountForm
+              intl={intl}
+              onSubmitDeleteAccount={values => handleSubmit(values)}
```
```suggestion
              onSubmitDeleteAccount={handleSubmit}
```

Amanda-Kauppinen committed:
Translation: Updated key from deleteAccountTabTitle -> accountManagementTabTitle

Amanda-Kauppinen committed:
ManageAccountPage: updated handleSubmit function call in DeleteAccountForm parameter

Amanda-Kauppinen committed:
Renamed tab 'Account management' > 'Manage account'

Amanda-Kauppinen committed:
Moved account delete function to api.js

Amanda-Kauppinen committed:
delete-account: reworking account deletion logic

Amanda-Kauppinen committed:
sdk: Added error message handling for ongoing transaction error message

Amanda-Kauppinen committed:
ManageAccountPage: passing deleteAccountError to form, fixing deleteAccount function chaining

Amanda-Kauppinen committed:
DeleteAccountForm: WIP adding error handling for UI

Amanda-Kauppinen committed:
DeleteAccountForm: Moved error handling to function

Amanda-Kauppinen committed:
delete-account: fixed transaction checking, WIP will improve

Amanda-Kauppinen committed:
ManageAccountPage.duck: Fixed invalid syntax

Amanda-Kauppinen committed:
ManageAccountPage: Removed unused params, passed required params to form

Amanda-Kauppinen committed:
DiscardDraftModal: Added dark red to focus and hover for submit

Amanda-Kauppinen committed:
DefaultAccountForm: Added focus/hover color to button, removed unused classes

Amanda-Kauppinen committed:
DeleteAccountForm: Fixed error handling for invalid password

Added password reset logic - WIP

Gnito commented on server/api-util/sdk.js:
```diff
@@ -96,6 +96,16 @@ exports.handleError = (res, error) => {
         statusText: error.message,
       })
       .end();
+  } else if (error.message == 'Account has ongoing transactions with Stripe states') {
```
Perhaps this could be formulated a bit differently. Perhaps something like this:

```js
  } else if (error.message == 'User has transactions on states that include incomplete payment processing') {
```

"Stripe states" sounds quite strange - and if we / customizers add other payment providers, it would be outdated.

Gnito commented on server/api/delete-account.js:
```diff
@@ -0,0 +1,73 @@
+    'state/delivered',
+    'state/changes-requested',
+    'state/completed',
+  ];
```
Perhaps these could be moved to the file-scope
(above this middleware definition `module.exports = (req, res) => {`)

That way it would be easier to follow what logic is actually inside it.

Gnito commented on server/api/delete-account.js:
```diff
@@ -0,0 +1,73 @@
+
+const hasOngoingTransactionsWithStripeRelatedStates = responses => {
+  // TODO: improve
+  return responses.some(map => Array.isArray(map?.data?.data) && map.data.data.length > 0);
```
Perhaps at least naming the "map?.data?.data;"
```js
  return responses.some(map => {
    const queryResults = map?.data?.data;
    return Array.isArray(queryResults) && queryResults.length > 0;
  });
```

Gnito commented on src/routing/routeConfiguration.js:
```diff
@@ -28,6 +28,7 @@ const PasswordChangePage = loadable(() => import(/* webpackChunkName: "PasswordC
 const PasswordRecoveryPage = loadable(() => import(/* webpackChunkName: "PasswordRecoveryPage" */ '../containers/PasswordRecoveryPage/PasswordRecoveryPage'));
 const PasswordResetPage = loadable(() => import(/* webpackChunkName: "PasswordResetPage" */ '../containers/PasswordResetPage/PasswordResetPage'));
 const PaymentMethodsPage = loadable(() => import(/* webpackChunkName: "PaymentMethodsPage" */ '../containers/PaymentMethodsPage/PaymentMethodsPage'));
+const ManageAccountPage = loadable(() => import(/* webpackChunkName: "ManageAccountPage" */ '../containers/ManageAccountPage/ManageAccountPage'));
```
Perhaps this could be moved next to ManageListingsPage (it seems this is mostly in alphabetical order. 
(It seems I failed to move MakeOfferPage after it was renamed as such after its previous name of InitiateNegotiationPage.

Gnito commented on src/containers/ManageAccountPage/DeleteAccountForm/DeleteAccountForm.js:
```diff
@@ -0,0 +1,213 @@
+            </div>
+            <div className={css.bottomWrapper}>
+              {// TODO: refine
+              passwordErrorText ? null : <ErrorMessage error={deleteAccountError} />}
```
Perhaps, the ErrorMessage component should render null if the error is something that's handled elsewhere.

Amanda-Kauppinen committed:
Various files: Renamed errors/variables 'stripe related payments' to 'incomplete payment processing'

Amanda-Kauppinen committed:
delete-account: Moved booking state definitions to file scope

Updated directory README to describe delete-account, with
changelog formatting update

Amanda-Kauppinen committed:
routeConfiguration: moved ManageAccountPage to be by ManageListingsPage

Amanda-Kauppinen committed:
Reworked transaction query response checking, added API error response handling

Amanda-Kauppinen committed:
DeleteAccountForm: Tidied error handling, cleaned up unused variables

Amanda-Kauppinen committed:
ManageAccountPage.duck: Removed unnecessary throw

Amanda-Kauppinen committed:
ManageAccountPage: Added page reload after successful account deletion

Amanda-Kauppinen committed:
DeleteAccountForm: Fixed incorrect
password handling and error message

Amanda-Kauppinen committed:
DeleteAccountForm.module.css: Added disabled

Amanda-Kauppinen committed:
Various code cleanup and adding comments

SariSaar committed:
Add navigation props for showing conditional links

SariSaar committed:
Update ManageAccountPage.duck.js to Redux toolkit

SariSaar committed:
Add parameter and error handling for Stripe deletion

SariSaar committed:
Update translations

SariSaar committed:
Redirect straight to landing page after account deletion

SariSaar committed:
Skip server-side error logging for incomplete transactions error

Gnito reviewed (approved):
LGTM

Gnito committed:
DeleteAccountForm: disable submit until currentPassword is given

Amanda-Kauppinen committed:
Updated changelog
